File: client/src/lib/watch-camera-pip.ts
/**
 * The presenter's camera, floating over their film.
 *
 * WHAT IT IS. A watch party's audience is seatless — nobody outside the room
 * holds a LiveKit seat — so a camera published into the room reaches the
 * seated participants and nobody on the playlist. The server therefore runs a
 * second, video-only egress (480p, or 360p) beside the ladder and states its playlist
 * as `LiveHlsStream.cameraHlsUrl`. This module is the viewer's half of that:
 * where the picture-in-picture sits, whether it is showing at all, and which
 * of the two pictures is on the stage. See `docs/WATCH_PARTY.md`, "The
 * presenter's camera, floating over the film".
 *
 * WHY IT IS A SEPARATE MODULE. Everything here is pure and everything here is
 * a rule a person will notice immediately if it is wrong: a webcam covering
 * the subtitles, a corner that forgets itself between films, a PiP that stays
 * up in fullscreen. `hls-watch-player.tsx` is a 1,500-line component whose
 * tests cannot reach past the first render, so a rule that lives inside it is
 * a rule nothing checks.
 *
 * THE STAGE AND THE CORNER ARE BOXES, NOT PLAYERS. A layout change moves
 * which `<video>` gets which class, and nothing else: neither hls.js instance is
 * re-attached, so nobody rebuffers to look at a webcam, and the control bar
 * stays where it is because it belongs to the stage rather than to a picture.
 */

export const CAMERA_PIP_CORNERS = [
  "top-left",
  "top-right",
  "bottom-left",
  "bottom-right",
] as const;

export type CameraPipCorner = (typeof CAMERA_PIP_CORNERS)[number];

/**
 * HOW THE VIEWER WANTS THE TWO PICTURES (2026-09-25), Rafael's four:
 *
 *  - `pip`: the default. The film on the stage, the webcam small in a corner.
 *  - `side`: the two next to each other; stacked, film on top, on a stage
 *    narrower than a phone held sideways (`CAMERA_SIDE_*_CLASS`).
 *  - `stream`: "hide webcam". The film alone. The camera's player is
 *    UNMOUNTED, so its playlist stops downloading and nothing decodes it,
 *    unless it is also carrying the presenter's voice ("separada"), which a
 *    hidden webcam must not silence: then it stays, drawn as the voice-only
 *    corner.
 *  - `camera`: "hide stream". The webcam on the stage, alone. The film's
 *    player keeps playing underneath, covered, because it is the one carrying
 *    the audio (the party's sound is mixed into it), and detaching it would
 *    be exactly the rebuffer this module exists to avoid: switching back has
 *    to be instant. Covered, not `display: none`, so no browser treats it as
 *    an offscreen video it may pause.
 */
export const CAMERA_LAYOUTS = ["pip", "side", "stream", "camera"] as const;

export type CameraLayout = (typeof CAMERA_LAYOUTS)[number];

export interface CameraPipPref {
  corner: CameraPipCorner;
  /** Which of the four. Remembered like the corner: a whole-party choice. */
  layout: CameraLayout;
}

/**
 * Bottom right, film on the stage.
 *
 * Bottom right because that is where every video call in the world puts the
 * small picture, and because the top of this player is where the audience
 * count, the leave button and the live badge already live. The corner is
 * movable precisely because "bottom right" is wrong for the one film whose
 * subtitles are burnt in down there.
 */
export const DEFAULT_CAMERA_PIP: CameraPipPref = {
  corner: "bottom-right",
  layout: "pip",
};

const STORAGE_KEY = "pqp:watch-camera-pip";

/**
 * A stored preference, defensively. An entry written before the layouts
 * existed carried `onStage` (the old swap, camera big and film small); it is
 * read as the default rather than as "hide stream", because opening a party
 * to no film at all is not what that person asked for.
 */
export function parseCameraPipPref(raw: unknown): CameraPipPref {
  if (!raw || typeof raw !== "object") {
    return DEFAULT_CAMERA_PIP;
  }
  const value = raw as Partial<Record<keyof CameraPipPref, unknown>>;
  return {
    corner: CAMERA_PIP_CORNERS.includes(value.corner as CameraPipCorner)
      ? (value.corner as CameraPipCorner)
      : DEFAULT_CAMERA_PIP.corner,
    layout: CAMERA_LAYOUTS.includes(value.layout as CameraLayout)
      ? (value.layout as CameraLayout)
      : DEFAULT_CAMERA_PIP.layout,
  };
}

/**
 * Per browser, and never a reason for the player not to render.
 *
 * Every read and write is wrapped: `localStorage` throws outright in a
 * thumbnail capture and in a browser set to block site data, and a watch party
 * that fails to draw because a corner preference could not be read would be a
 * spectacular way to lose a film.
 */
export function readCameraPipPref(): CameraPipPref {
  try {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    return raw ? parseCameraPipPref(JSON.parse(raw)) : DEFAULT_CAMERA_PIP;
  } catch {
    return DEFAULT_CAMERA_PIP;
  }
}

export function writeCameraPipPref(next: CameraPipPref): void {
  try {
    window.localStorage.setItem(STORAGE_KEY, JSON.stringify(next));
  } catch {
    // A per-viewer convenience, not state anything depends on.
  }
}

/** The next corner, clockwise. Four presses is where you started. */
export function nextCameraPipCorner(corner: CameraPipCorner): CameraPipCorner {
  const index = CAMERA_PIP_CORNERS.indexOf(corner);
  return CAMERA_PIP_CORNERS[(index + 1) % CAMERA_PIP_CORNERS.length]!;
}

/**
 * Whether the viewer is offered the layout picker at all: only for a camera
 * that is a picture. No camera playlist is today's stage exactly, and the
 * audio-only "separada" shape (`cameraHasVideo` false) has nothing to lay out.
 */
export function cameraLayoutOffered(input: {
  cameraSrc: string | null | undefined;
  cameraHasVideo: boolean;
  cinema: boolean;
}): boolean {
  return Boolean(input.cameraSrc) && input.cameraHasVideo && input.cinema;
}

/** The layout in force: the viewer's, when there is a picture to lay out. */
export function effectiveCameraLayout(input: {
  pref: CameraPipPref;
  cameraHasVideo: boolean;
}): CameraLayout {
  return input.cameraHasVideo ? input.pref.layout : "pip";
}

/**
 * Whether the camera player exists on this stage at all.
 *
 *  - A `cameraHlsUrl`: the server is running a camera transcode. Absent for a
 *    host with no webcam on, for a box that refused it on budget, and for
 *    `LIVE_HLS_CAMERA=false`.
 *  - Cinema layout only. A webcam inside a grid tile (`tile`) is a picture in
 *    a picture in a picture, and the docked mini player (`mini`) is a 240px
 *    box with room for the film and almost nothing else.
 *  - **Not "hide webcam"**, unless the playlist carries the presenter's voice.
 *    Unmounted rather than hidden, so a camera nobody asked to see costs no
 *    download and no decode.
 *
 * FULLSCREEN NO LONGER UNMOUNTS IT (2026-09-25). It used to, when a corner
 * was the only way to show a camera and fullscreen meant "the film and
 * nothing else". With the picker, the viewer says what fullscreen shows:
 * whatever layout they chose, including "hide webcam" for the film alone.
 */
export function cameraPipMounted(input: {
  cameraSrc: string | null | undefined;
  cinema: boolean;
  layout: CameraLayout;
  hasVoiceAudio: boolean;
}): boolean {
  if (!input.cameraSrc || !input.cinema) {
    return false;
  }
  return input.layout !== "stream" || input.hasVoiceAudio;
}

/** The full-bleed picture. */
export const CAMERA_PIP_STAGE_CLASS = "absolute inset-0 h-full w-full";

/**
 * The camera as the whole stage ("hide stream"): full bleed, opaque, above the
 * film it covers and below the holding screens (`STAGE_LAYER.state`, z-10),
 * so a film that stalls still says so over the face.
 */
export const CAMERA_STAGE_CLASS = "absolute inset-0 z-[5] h-full w-full bg-black";

/**
 * Side by side, halves of the stage. Measured against the PLAYER's width
 * (`@container/watch` on its root), not the window's: the same stage is a
 * full desktop pane, a split beside the chat and a phone, and it is the box
 * that decides whether two 16:9 pictures fit next to each other. Under 36rem
 * (a phone held upright) they stack, film on top.
 */
export const CAMERA_SIDE_FILM_CLASS =
  "absolute inset-x-0 top-0 h-1/2 w-full @xl/watch:inset-y-0 @xl/watch:right-auto @xl/watch:h-full @xl/watch:w-1/2";
export const CAMERA_SIDE_CAMERA_CLASS =
  "absolute inset-x-0 bottom-0 h-1/2 w-full bg-black @xl/watch:inset-y-0 @xl/watch:left-auto @xl/watch:h-full @xl/watch:w-1/2";

/**
 * The corner box.
 *
 * Percentages with a floor and a ceiling: 24 % of a 1440px pane is a 345px
 * webcam, which is about right, and the same 24 % of a phone-width pane is
 * 96px, which is a smudge. `min-w` and `max-w` are what keep it a face at
 * both ends.
 */
const CAMERA_PIP_FRAME_CLASS =
  "absolute aspect-video w-[24%] min-w-[128px] max-w-[280px]";

/**
 * The look, which only the picture wants and the click target must not have.
 *
 * `z-20` puts the picture over the film and under the chrome (z-50), so the
 * control bar is never behind a webcam. The corner control is given z-30 by
 * its caller, between the two.
 */
const CAMERA_PIP_SKIN_CLASS =
  "z-20 overflow-hidden rounded-[var(--radius-card)] border border-paper/25 bg-black shadow-lg";

const CORNER_CLASS: Record<CameraPipCorner, string> = {
  // Clear of the chrome's own gradients: the top bar carries the audience
  // count and the actions, the bottom one the whole control bar.
  "top-left": "left-3 top-14",
  "top-right": "right-3 top-14",
  "bottom-left": "bottom-20 left-3",
  "bottom-right": "bottom-20 right-3",
};

/**
 * Where the corner is and how big, with no appearance of its own.
 *
 * Split from the skin because the corner control sits in exactly this box
 * and must NOT carry the background, the border or the rounding: a control
 * that is sometimes opaque is a control that sometimes hides the picture it
 * is controlling.
 */
export function cameraPipFrameClass(corner: CameraPipCorner): string {
  return `${CAMERA_PIP_FRAME_CLASS} ${CORNER_CLASS[corner]}`;
}

export function cameraPipCornerClass(corner: CameraPipCorner): string {
  return `${cameraPipFrameClass(corner)} ${CAMERA_PIP_SKIN_CLASS}`;
}

export interface CameraPipBoxes {
  /** Classes for the film's `<video>`. */
  film: string;
  /** Classes for the camera's box, or null when it is not mounted. */
  camera: string | null;
  /**
   * The box the corner control sits in, or null when there is no corner to
   * move (no frame yet, or a layout with no corner in it).
   */
  corner: string | null;
  /**
   * How the camera fills its box: `cover` in the corner (a face cropped to
   * 16:9 is still a face, and a letterboxed thumbnail is mostly black),
   * `contain` wherever it is one of the two main pictures.
   */
  cameraFit: "cover" | "contain";
  /**
   * The camera is mounted only for the presenter's voice ("hide webcam" in
   * "separada"): draw it as the voice-only corner, not as a picture.
   */
  cameraVoiceOnly: boolean;
}

/**
 * Which picture gets which box.
 *
 * All four layouts in one pure function, so every state reads at once, and
 * every one of them is a CLASS change: neither hls.js instance is ever
 * re-attached by a layout switch, so nobody rebuffers to look at a webcam and
 * the control bar stays where it is because it belongs to the stage.
 *
 * `mounted` AND `hasFrame` ARE DIFFERENT QUESTIONS, and collapsing them is a
 * deadlock: a camera that is not rendered never decodes a frame, so gating the
 * element on having one means it never gets one. Mounted is "the player
 * exists and is loading"; a frame is what makes it worth looking at. Between
 * the two it is `invisible` in the corner and the film keeps the whole stage
 * whatever the layout, so "side by side" or "hide stream" never opens on a
 * black half or a black stage while the camera loads.
 */
export function cameraPipBoxes(input: {
  mounted: boolean;
  hasFrame: boolean;
  pref: CameraPipPref;
  layout: CameraLayout;
}): CameraPipBoxes {
  const none: CameraPipBoxes = {
    film: CAMERA_PIP_STAGE_CLASS,
    camera: null,
    corner: null,
    cameraFit: "cover",
    cameraVoiceOnly: false,
  };
  if (!input.mounted) {
    return none;
  }
  const corner = cameraPipCornerClass(input.pref.corner);
  const frame = cameraPipFrameClass(input.pref.corner);
  if (input.layout === "stream") {
    // Mounted here only for the voice: the corner, drawn as the voice.
    return { ...none, camera: corner, cameraVoiceOnly: true };
  }
  if (!input.hasFrame) {
    // Loading, and drawing nothing. A camera that never produces a frame must
    // cost the film nothing, not even a rectangle: a rectangle is exactly
    // what a viewer reads as the feature being broken.
    return { ...none, camera: `${corner} invisible` };
  }
  switch (input.layout) {
    case "side":
      return {
        ...none,
        film: CAMERA_SIDE_FILM_CLASS,
        camera: CAMERA_SIDE_CAMERA_CLASS,
        cameraFit: "contain",
      };
    case "camera":
      return {
        ...none,
        camera: CAMERA_STAGE_CLASS,
        cameraFit: "contain",
      };
    default:
      return { ...none, camera: corner, corner: frame };
  }
}

File: ios/pqp/Sources/Voice/WatchCameraPip.swift
import AVKit
import SwiftUI

/**
 THE PRESENTER'S CAMERA, FLOATING OVER THEIR FILM.

 Mirrors `client/src/lib/watch-camera-pip.ts` and `watch-camera-pip.tsx`. A
 watch party's audience is seatless, so a camera published into the room
 reaches nobody on the playlist; the server runs a second, video-only egress
 beside the ladder and states its playlist as `LiveHlsStream.cameraHlsUrl`.
 This file is the iOS half: where the picture-in-picture sits, whether it is
 showing at all, and which of the two pictures owns the stage.

 DELIBERATELY MUCH SIMPLER THAN THE FILM'S PLAYER (`WatchStageView`,
 `WatchVideoSurface`). The camera carries one rendition, no ladder, no
 fullscreen, no system Picture-in-Picture of its own, and a failure whose
 correct response is to disappear rather than to tell anybody anything. See
 `WatchCameraSurface`.
 */

// MARK: - Layout

/// Mirrors `CAMERA_LAYOUTS` in `client/src/lib/watch-camera-pip.ts`, Rafael's
/// four: the default corner, side by side, hide the camera, hide the film.
enum CameraLayout: String, CaseIterable, Codable, Equatable, Sendable {
    /// The default. The film on the stage, the camera small in a corner.
    case pip
    /// The two next to each other (stacked in portrait, film on top; side by
    /// side once the stage is wider than it is tall).
    case side
    /// "Hide camera". The film alone. The camera's player is unmounted,
    /// UNLESS it also carries the presenter's voice, which a hidden webcam
    /// must not silence: then it stays, drawn as a voice-only corner.
    case stream
    /// "Hide stream". The camera on the stage, alone. The film keeps playing
    /// underneath, covered rather than torn down, because it is the one
    /// carrying the party's audio and switching back has to be instant.
    case camera
}

/// Mirrors `CAMERA_PIP_CORNERS`. `Alignment` is what `ZStack` wants; the raw
/// cases are what gets persisted and compared.
enum CameraPipCorner: String, CaseIterable, Codable, Equatable, Sendable {
    case topLeading, topTrailing, bottomLeading, bottomTrailing

    var alignment: Alignment {
        switch self {
        case .topLeading: return .topLeading
        case .topTrailing: return .topTrailing
        case .bottomLeading: return .bottomLeading
        case .bottomTrailing: return .bottomTrailing
        }
    }

    /// The corner closest to a drag's release point inside a stage of the
    /// given size. Pure so a drag gesture's end can be asserted on without a
    /// running view.
    static func nearest(to point: CGPoint, in bounds: CGSize) -> CameraPipCorner {
        let leading = point.x < bounds.width / 2
        let top = point.y < bounds.height / 2
        switch (top, leading) {
        case (true, true): return .topLeading
        case (true, false): return .topTrailing
        case (false, true): return .bottomLeading
        case (false, false): return .bottomTrailing
        }
    }

    /// The next corner, clockwise. Four presses is where you started.
    /// Mirrors `nextCameraPipCorner`; VoiceOver's "move to another corner"
    /// action uses this, since a drag gesture is not reachable that way.
    func clockwise() -> CameraPipCorner {
        switch self {
        case .topLeading: return .topTrailing
        case .topTrailing: return .bottomTrailing
        case .bottomTrailing: return .bottomLeading
        case .bottomLeading: return .topLeading
        }
    }
}

/// What the viewer asked for, remembered per phone. Bottom trailing (the
/// corner every video call on the platform already uses) and `pip`, the
/// default, film on the stage.
///
/// DELIBERATELY NOT `Codable` ITSELF. The standard library gives any type
/// that is both `Codable` and `RawRepresentable` with an `Encodable`
/// `RawValue` a DEFAULT `encode(to:)`/`init(from:)` built out of
/// `rawValue`/`init?(rawValue:)` -- and this type's `rawValue` is built out
/// of `JSONEncoder`, so the two defaults call each other. Confirmed the hard
/// way: a `SIGSEGV` stack overflow pinned on `CameraPipPref.rawValue.getter`
/// calling itself through `JSONEncoder.encode`, every time a test actually
/// exercised the round trip. `StoredCameraPipPref` below is the `Codable`
/// half instead, so the two conformances never meet.
struct CameraPipPref: Equatable, Sendable {
    var corner: CameraPipCorner
    var layout: CameraLayout

    static let `default` = CameraPipPref(corner: .bottomTrailing, layout: .pip)
}

/// The JSON shape alone, with no `RawRepresentable` in sight.
private struct StoredCameraPipPref: Codable {
    var corner: CameraPipCorner
    var layout: CameraLayout
}

/**
 `@AppStorage`-backed via JSON, the same shape `readCameraPipPref` /
 `writeCameraPipPref` give the web's `localStorage`. Defensive by
 construction: a value that fails to decode (nothing stored yet, a future
 format) reads as `.default` rather than crashing the stage over a
 preference nobody would notice missing.
 */
extension CameraPipPref: RawRepresentable {
    init?(rawValue: String) {
        guard let data = rawValue.data(using: .utf8),
              let decoded = try? JSONDecoder().decode(StoredCameraPipPref.self, from: data)
        else {
            self = .default
            return
        }
        self.init(corner: decoded.corner, layout: decoded.layout)
    }

    var rawValue: String {
        let stored = StoredCameraPipPref(corner: corner, layout: layout)
        guard let data = try? JSONEncoder().encode(stored),
              let string = String(data: data, encoding: .utf8)
        else { return "{}" }
        return string
    }
}

/// Whether the viewer is offered the layout menu at all: only for a camera
/// that is actually a picture. Mirrors `cameraLayoutOffered`.
func cameraLayoutOffered(cameraSrc: String?, cameraHasVideo: Bool) -> Bool {
    cameraSrc != nil && cameraHasVideo
}

/// The layout in force. A camera with no picture (the audio-only "separada"
/// shape) has nothing to lay out and always reads as the corner box, which is
/// also where the voice-only indicator lives. Mirrors `effectiveCameraLayout`.
func effectiveCameraLayout(pref: CameraPipPref, cameraHasVideo: Bool) -> CameraLayout {
    cameraHasVideo ? pref.layout : .pip
}

/**
 THE URL WORTH A LIVE CONNECTION FOR, GIVEN WHAT THE VIEWER CAN ACTUALLY SEE
 OR HEAR RIGHT NOW.

 "Hide camera" (`.stream`) with nothing to hear either is nothing worth
 streaming: `nil` here is what makes `WatchCameraStreamSwap` detach a camera
 the viewer explicitly hid, instead of leaving it running invisibly for as
 long as that layout stays picked (Farol review, PR 833). A camera that DOES
 carry the presenter's voice keeps streaming even hidden -- the corner still
 shows the voice-only indicator for it, in every layout but this one.
 */
func cameraUrlWorthStreaming(
    cameraHlsUrl: String?,
    hasVoiceAudio: Bool,
    layoutOffered: Bool,
    layout: CameraLayout
) -> String? {
    let hiddenAndSilent = layoutOffered && layout == .stream && !hasVoiceAudio
    return hiddenAndSilent ? nil : cameraHlsUrl
}

// MARK: - The swap rule

/// What the camera player currently holds, so the swap rule can tell a
/// restamp from a restart. Keyed on the playlist's own path rather than on
/// `LiveHlsStream.startedAt`: the camera shares its parent broadcast's
/// session and never mints an identity of its own on the wire.
struct CameraAttachedStream: Equatable, Sendable {
    let sessionKey: String
    let attachedAt: Date
}

enum CameraPlayerMove: Equatable, Sendable {
    case keep
    case attach(hlsUrl: String)
    case detach
}

/// The camera's own `hlsSessionKey`: everything before the `?`, which is
/// where the server's per-viewer, per-keyframe `?t=` restamp lives. Mirrors
/// `hlsSessionKey` in `client/src/lib/hls-playback.ts`.
func cameraSessionKey(_ url: String) -> String {
    guard let query = url.firstIndex(of: "?") else { return url }
    return String(url[url.startIndex..<query])
}

/**
 WHEN TO HAND THE CAMERA'S `AVPlayer` A DIFFERENT URL, AND, MOSTLY, WHEN NOT
 TO.

 Same shape as `WatchStreamSwap`, and the same bug it exists to avoid: the
 server restamps `cameraHlsUrl`'s `?t=` on the same audience-keyframe clock as
 the film's, about every 30 seconds, for a camera that has not moved at all.
 Rebuilding the player on every restamp would tear it down, drop the buffer
 and blank the corner on that cadence for the whole party.

 A camera with no picture and no voice track is not drawn at all
 (`.detach`); the caller decides whether to even ask (see
 `cameraLayoutOffered`), this only decides the URL once it has.
 */
enum WatchCameraStreamSwap {
    /// Same margin as `WatchStreamSwap.renewAfter`: the viewer token lives an
    /// hour, and this refreshes comfortably inside it rather than at the
    /// edge of expiry.
    static let renewAfter: TimeInterval = 50 * 60

    static func next(
        attached: CameraAttachedStream?,
        latestUrl: String?,
        hasVideo: Bool,
        hasVoiceAudio: Bool,
        failed: Bool,
        now: Date
    ) -> CameraPlayerMove {
        guard let latestUrl, hasVideo || hasVoiceAudio else { return .detach }
        let key = cameraSessionKey(latestUrl)
        guard let attached else { return .attach(hlsUrl: latestUrl) }
        if attached.sessionKey != key { return .attach(hlsUrl: latestUrl) }
        if failed { return .attach(hlsUrl: latestUrl) }
        if now.timeIntervalSince(attached.attachedAt) >= renewAfter {
            return .attach(hlsUrl: latestUrl)
        }
        return .keep
    }
}

// MARK: - The surface

/// One `AVPlayerLayer`, drawn plain. No PiP hookup, no fullscreen, no
/// autoresizing dance across a theater move: the camera never leaves its box.
final class CameraPlayerCanvas: UIView {
    override class var layerClass: AnyClass { AVPlayerLayer.self }

    var playerLayer: AVPlayerLayer { layer as! AVPlayerLayer }

    var player: AVPlayer? {
        get { playerLayer.player }
        set { playerLayer.player = newValue }
    }
}

/// The camera's rectangle. `fit` is `.resizeAspectFill` in the corner box
/// (a face fills a small circle better than it letterboxes in one) and
/// `.resizeAspect` when the camera IS the stage (`side`, `camera` layouts),
/// mirroring the web's `cameraPipBoxes` fit choice.
struct WatchCameraSurface: UIViewRepresentable {
    let player: AVPlayer?
    var fit: AVLayerVideoGravity = .resizeAspectFill

    func makeUIView(context: Context) -> CameraPlayerCanvas {
        let canvas = CameraPlayerCanvas()
        canvas.backgroundColor = .black
        canvas.playerLayer.videoGravity = fit
        canvas.player = player
        return canvas
    }

    func updateUIView(_ canvas: CameraPlayerCanvas, context: Context) {
        canvas.playerLayer.videoGravity = fit
        if canvas.player !== player { canvas.player = player }
    }
}
